gui/src/2D/controls/structuredTextBlock.ts

```diff
@@ -0,0 +1,779 @@
+import { Observable } from "babylonjs/Misc/observable";
+import { Measure } from "../measure";
+import { ValueAndUnit } from "../valueAndUnit";
+import { Control } from "./control";
+import { TextWrapping } from "./textBlock";
+import { RegisterClass } from "babylonjs/Misc/typeStore";
+import { Nullable } from "babylonjs/types";
+import { serialize } from 'babylonjs/Misc/decorators';
+import { ICanvasRenderingContext , ICanvasGradient } from 'babylonjs/Engines/ICanvas';
+import { Engine } from 'babylonjs/Engines/engine';
+
+export interface StructuredTextPart {
+  text: string;
+
+  color?: string | ICanvasGradient;
+
+  underline?: boolean;
+  lineThrough?: boolean;
+
+  // For instance, font size and family is not updatable, the whole StructuredTextBlock shares the same size and family (not useful and it introduces complexity)
+  fontStyle?: string;
+  fontWeight?: string;
+
+  outlineWidth?: number;
+  outlineColor?: string;
+
+  shadowColor?: string;
+  shadowBlur?: number;
+  shadowOffsetX?: number;
+  shadowOffsetY?: number;
+
+  // Computed part width
+  width?: number;
+}
+
+export type StructuredText = Array<StructuredTextPart>;
+
+type StructuredTextLine = {
+  parts: StructuredText;
+
+  // Computed line width
+  width?: number;
+};
+type StructuredTextLines = Array<StructuredTextLine>;
+
+// Mostly the same than StructuredTextPart, but nothing is optional here,
+// this is the attributes about to be sent to the canvas context.
+interface TextPartAttributes {
+  color: string | ICanvasGradient;
+
+  underline: boolean;
+  lineThrough: boolean;
+
+  // For instance, font size and family is not updatable, the whole StructuredTextBlock shares the same size and family (not useful and it introduces complexity)
+  fontStyle: string;
+  fontWeight: string;
+
+  outlineWidth: number;
+  outlineColor: string;
+
+  shadowColor: string;
+  shadowBlur: number;
+  shadowOffsetX: number;
+  shadowOffsetY: number;
+}
+
+/**
+ * Class used to create structured text block control
+ */
+//export class StructuredTextBlock extends TextBlock {
+export class StructuredTextBlock extends Control {
```

Comment:
up @cronvel

Reply:
@deltakosh 
> About the `TextBlock`'s inheritance, there isn't much to inherit from, only one method is preserved from the `TextBlock` class and it may change too. It would get in the way more than it helps. Also there is no property `.text` on `StructuredTextBlock`. It would just create more confusion than it solves, especially if more feature are added in the future (Markdown parser, hypertext-like button/hover/tooltip, etc).

Reply:
I understand your point. That being said it will help users as they could have a function taking a TextBlock that cn also work with StructuredTExtBlock

Does it make sense?

Reply:
@deltakosh The problem is that there are already few incompatibilities between both classes, and it can only diverge more in the future.
There is even some **public properties that just doesn't exist anymore** on StructuredTextBlock (e.g. there is no `.text`).